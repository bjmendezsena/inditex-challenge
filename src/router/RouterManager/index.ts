export * from './RouteName';
export * from './RoutePath';
