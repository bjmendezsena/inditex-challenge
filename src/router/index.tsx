import { RouterProvider, createBrowserRouter } from "react-router-dom";
import routes from "./Routes";

export const router = createBrowserRouter(routes);

const AppRouter = () => <RouterProvider router={router} />;

export default AppRouter;
