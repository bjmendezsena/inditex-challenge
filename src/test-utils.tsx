import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
