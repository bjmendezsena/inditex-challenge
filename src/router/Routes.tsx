import { RouteObject, Navigate } from "react-router-dom";
import * as Views from "../views";
import { MainLayout, PodcastLayout } from "../layouts";
import { RoutePath } from "./RouterManager";

const routes: RouteObject[] = [
  {
    element: <MainLayout />,
    children: [
      {
        path: RoutePath.PodcastList,
        element: <Views.PodcastListPage />,
      },
      {
        element: <PodcastLayout />,
        children: [
          {
            path: RoutePath.PodcastDetails,
            element: <Views.PodcastDetails />,
          },
          {
            path: RoutePath.EpisodeDetails,
            element: <Views.EpisodeDetails />,
          },
        ],
      },
      {
        path: "*",
        element: <Navigate to={RoutePath.PodcastList} replace />,
      },
    ],
  },
];

export default routes;
