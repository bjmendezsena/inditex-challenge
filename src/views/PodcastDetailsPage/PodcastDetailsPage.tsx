import { useEffect, useMemo } from "react";
import { createColumnHelper } from "@tanstack/react-table";
import { useParams, Link as RRDLink } from "react-router-dom";
import { format } from "date-fns";
import {
  Card,
  Container,
  CardBody,
  Flex,
  Heading,
  Skeleton,
  Link,
} from "@chakra-ui/react";
import { Text, Table } from "../../components";
import { usePodcastDetails } from "../../hooks";
import { RouterManager, RouteName } from "../../router";
import { Result } from "../../interfaces/podcastDetails";

const DURATION_FORMAT = "mm:ss";
const DATE_FORMAT = "dd/MM/yyyy";

interface PodcastTableData {
  trackName: string;
  releaseDate: string;
  trackTimeMillis: number;
  collectionId: number;
  trackId: number;
  raw: Result;
}

const columnHelper = createColumnHelper<PodcastTableData>();

const columns = [
  columnHelper.accessor("trackName", {
    header: "Title",
    size: 400,
    cell: ({ getValue, row }) => {
      const { original } = row;
      console.log({ original });
      return (
        <Link
          data-testid={`episode-${original.trackId}`}
          as={RRDLink}
          to={RouterManager.getUrl(RouteName.EpisodeDetails, {
            podcastId: original.collectionId,
            episodeId: original.trackId,
          })}
          sx={{
            fontWeight: "regular",
          }}
        >
          {getValue()}
        </Link>
      );
    },
  }),
  columnHelper.accessor("releaseDate", {
    header: "Date",
    cell: ({ getValue }) => {
      return <Text>{format(new Date(getValue()), DATE_FORMAT)}</Text>;
    },
  }),
  columnHelper.accessor("trackTimeMillis", {
    header: "Duration",
    cell: ({ getValue }) => (
      <Text>{format(new Date(getValue()), DURATION_FORMAT)}</Text>
    ),
  }),
];

export const PodcastDetailsPage = () => {
  const { podcastId } = useParams<{ podcastId: string }>();

  useEffect(() => {
    !podcastId &&
      RouterManager.to(RouteName.PodcastList, {
        replace: true,
      });
  });

  const { data, podcastInfo, isLoading } = usePodcastDetails(podcastId);

  const sourceData: PodcastTableData[] = useMemo(() => {
    if (!data) return [];
    return data.results
      .filter((item) => !item.artistName)
      .map((item: Result) => ({
        trackName: item.trackName,
        releaseDate: item.releaseDate,
        trackTimeMillis: item.trackTimeMillis,
        collectionId: item.collectionId,
        trackId: item.trackId,
        raw: item,
      }));
  }, [data]);

  return (
    <Container maxW='container.md'>
      <Card
        sx={{
          width: "full",
          maxWidth: "container.md",
        }}
      >
        <CardBody
          sx={{
            minHeight: 0,
            height: "50px",
            pt: 2,
          }}
        >
          <Flex
            sx={{
              width: "full",
              alignItems: "center",
            }}
          >
            {isLoading ? (
              <Skeleton h={4} w={200} />
            ) : (
              <Heading>Episodes: {podcastInfo?.trackCount}</Heading>
            )}
          </Flex>
        </CardBody>
      </Card>
      <Card
        sx={{
          width: "full",
          maxWidth: "container.md",
        }}
      >
        <CardBody
          sx={{
            px: 4,
            display: "block",
            maxHeight: "500px",
          }}
        >
          <Table
            isLoading={isLoading}
            name='episode'
            columns={columns}
            data={sourceData}
          />
        </CardBody>
      </Card>
    </Container>
  );
};
