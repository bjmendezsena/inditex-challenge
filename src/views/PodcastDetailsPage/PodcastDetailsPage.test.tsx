import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { matchPath } from "react-router-dom";
import { render } from "../../test-utils";
import { RouterManager, RouteName } from "../../router";
import podcastDetailsJson from "../../dataset/podcastDetails.json";
import { PodcastDetailsPage } from "./PodcastDetailsPage";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    podcastId: "199",
  }),
}));

describe(`<${PodcastDetailsPage.name}/>`, () => {
  const history = createMemoryHistory();
  const factoryComponent = () =>
    render(<PodcastDetailsPage />, {
      history,
    });

  it("should render without crashing", () => {
    const { container } = factoryComponent();
    expect(container).toMatchSnapshot();
  });

  it("Should navigate to episode details page when click on episode name", async () => {
    factoryComponent();
    const episode = podcastDetailsJson.results[1];
    userEvent.click(screen.getByTestId(`episode-${episode.trackId}`));
    expect(
      matchPath(
        RouterManager.getUrl(RouteName.EpisodeDetails, {
          podcastId: episode.collectionId,
          episodeId: episode.trackId,
        }),
        history.location.pathname
      )
    ).toBeTruthy();
  });
});
