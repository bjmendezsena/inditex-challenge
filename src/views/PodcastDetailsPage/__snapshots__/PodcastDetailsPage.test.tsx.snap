// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PodcastDetailsPage/> should render without crashing 1`] = `
<div>
  <div
    class="chakra-container css-up7ra3"
  >
    <div
      class="chakra-card css-1w0i8ar"
    >
      <div
        class="chakra-card__body css-1wuju2"
      >
        <div
          class="css-uj3782"
        >
          <div
            class="chakra-skeleton css-yxcjy8"
          />
        </div>
      </div>
    </div>
    <div
      class="chakra-card css-1w0i8ar"
    >
      <div
        class="chakra-card__body css-1i7w1qe"
      >
        <div
          class="css-0"
        >
          <div
            class="css-t1ecuf"
          >
            <table
              class="chakra-table css-1td08nr"
              data-testid="table-episode"
            >
              <tbody
                class="css-tdnrhj"
              >
                <tr
                  class="css-bxak8j"
                >
                  <th
                    class="css-h4tswt"
                  >
                    Title
                  </th>
                  <th
                    class="css-223udn"
                  >
                    Date
                  </th>
                  <th
                    class="css-223udn"
                  >
                    Duration
                  </th>
                </tr>
                <tr
                  class="css-11j4jqy"
                  data-testid="row-0"
                >
                  <td
                    class="css-l132m4"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                </tr>
                <tr
                  class="css-11j4jqy"
                  data-testid="row-1"
                >
                  <td
                    class="css-l132m4"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                </tr>
                <tr
                  class="css-11j4jqy"
                  data-testid="row-2"
                >
                  <td
                    class="css-l132m4"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                </tr>
                <tr
                  class="css-11j4jqy"
                  data-testid="row-3"
                >
                  <td
                    class="css-l132m4"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                </tr>
                <tr
                  class="css-11j4jqy"
                  data-testid="row-4"
                >
                  <td
                    class="css-l132m4"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                  <td
                    class="css-175gdrf"
                    tabindex="0"
                  >
                    <div
                      class="chakra-skeleton css-yixgko"
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
